topdown/providers.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	return nil
+}
+
+func builtinAWSSigV4SignReq(ctx BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Request object.
+	reqObj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// AWS SigV4 config info object.
+	awsConfigObj, err := builtins.ObjectOperand(operands[1].Value, 1)
+	if err != nil {
+		return err
+	}
+	// Make sure our required keys exist!
+	err = validateAWSAuthParameters(awsConfigObj)
+	if err != nil {
+		return err
+	}
+	service := stringFromTerm(awsConfigObj.Get(ast.StringTerm("aws_service")))
+	awsCreds := providers.AWSCredentialsFromObject(awsConfigObj)
+
+	// Timestamp for signing.
+	var signingTimestamp time.Time
+	timestamp, err := builtins.NumberOperand(operands[2].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	ts, ok := timestamp.Int64()
+	if !ok {
+		return builtins.NewOperandErr(3, "could not convert time_ns value into a unix timestamp")
+	}
+
+	signingTimestamp = time.Unix(0, ts)
+	if err != nil {
+		return err
+	}
+
+	// Make sure our required keys exist!
+	// This check is stricter than required, but better to break here than downstream.
+	_, err = validateHTTPRequestOperand(operands[0], 1)
+	if err != nil {
+		return err
+	}
+
+	// Prepare required fields from the HTTP request object.
+	var theURL *url.URL
+	var method string
+	reqURL := reqObj.Get(ast.StringTerm("url"))
+	reqMethod := reqObj.Get(ast.StringTerm("method"))
+
+	headers := ast.NewObject()
+	headersTerm := reqObj.Get(ast.StringTerm("headers"))
+	if headersTerm != nil {
+		var ok bool
+		headers, ok = headersTerm.Value.(ast.Object)
+		if !ok {
+			return builtins.NewOperandTypeErr(0, headersTerm.Value, "object")
+		}
+	}
+
+	// Check types on the request parameters.
+	invalidParameters := ast.NewSet()
+	if _, ok := reqURL.Value.(ast.String); !ok {
+		invalidParameters.Add(ast.StringTerm("url"))
+	}
+	if _, ok := reqMethod.Value.(ast.String); !ok {
+		invalidParameters.Add(ast.StringTerm("method"))
+	}
+	if invalidParameters.Len() > 0 {
+		return builtins.NewOperandErr(1, "invalid values for required request parameters(s): %v", invalidParameters)
+	}
+
+	theURL, err = url.Parse(stringFromTerm(reqURL))
+	if err != nil {
+		return err
+	}
+	method = stringFromTerm(reqMethod)
+
+	bodyTerm := reqObj.Get(ast.StringTerm("body"))
+	rawBodyTerm := reqObj.Get(ast.StringTerm("raw_body"))
+	body, err := getReqBodyBytes(bodyTerm, rawBodyTerm)
+	if err != nil {
+		return err
+	}
+
+	// Sign the request object's headers, and reconstruct the headers map.
+	signedHeadersMap := providers.AWSSignV4(objectToMap(headers), method, theURL, body, service, awsCreds, signingTimestamp)
+	signedHeadersObj := ast.NewObject()
+	for k, v := range signedHeadersMap {
+		signedHeadersObj.Insert(ast.StringTerm(k), ast.StringTerm(strings.Join(v, ",")))
+	}
+
+	// Create new request object with updated headers.
+	out := reqObj.Copy()
+	out.Insert(ast.StringTerm("headers"), ast.NewTerm(signedHeadersObj))
```

Comment:
Fun fact: this would have blown up when doing `sign_req(data.some.req, ...)` without #5394. It's unlikely, but still. :D